core/src/UserCredential.cs

```diff
@@ -0,0 +1,100 @@
+//------------------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System.Runtime.InteropServices;
+using System.Security;
+
+namespace Microsoft.Identity.Core
+{
+    internal enum UserAuthType
+    {
+        IntegratedAuth,
+        UsernamePassword
+    }
+
+    /// <summary>
+    /// Credential used for integrated authentication on domain-joined machines.
+    /// The implementation is  here, and the ADAL UserCredential class will subclass this.
+    /// </summary>
+    public class UserCredential
```

Comment:
@henrik-me asked earlier in [another comment (now hidden by default)](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1118#discussion_r201540607) that, "this class should be internal?"

I attempted `internal` at the beginning too, and then I encountered several initially confusing build error:

>Error	CS0060	Inconsistent accessibility: base class 'UserCredential' is less accessible than class 'UserCredential'
>Error	CS0060	Inconsistent accessibility: base class 'UserCredential' is less accessible than class 'UserPasswordCred'

And that was because some of the subclass like same name `UserCredential` in ADAL has to be public, so now we hit a violation that "base class is less accessible than sub-class". It took me quite a while to really understand what happened. I ended up spending some extra time on writing a [StackOverflow answer for that](https://stackoverflow.com/questions/12135711/inconsistent-accessability-base-class-is-less-accessible-than-class/50477805#50477805), because, to me, if I could not clearly express something in written, that means I did not understand it good enough.

Reply:
This is violating the rule of not exposing anything from core. please move out. if you can't get this to work with inheritance then composition should be possible instead.

---
In reply to: [203918722](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1118#discussion_r203918722) [](ancestors = 203918722)

Reply:
Changed to internal in recent commit.

---
In reply to: [205346297](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1118#discussion_r205346297) [](ancestors = 205346297,203918722)